include/seqan3/alphabet/detail/hash.hpp

```diff
@@ -0,0 +1,97 @@
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Enrico Seiler <enrico.seiler AT fu-berlin.de>
+ * \brief Provides overloads for std::hash.
+ */
+
+#pragma once
+
+#include <range/v3/numeric/accumulate.hpp>
+
+#include <seqan3/core/metafunction/range.hpp>
+#include <seqan3/std/ranges>
+
+namespace std
+{
+/*!\brief Struct for hashing a character.
+ * \ingroup alphabet
+ * \tparam alphabet_t The type of character to hash; Must model seqan3::semi_alphabet_concept.
+ */
+template <seqan3::semi_alphabet_concept alphabet_t>
+struct hash<alphabet_t>
+{
+    /*!\brief Compute the hash for a character.
+     * \ingroup alphabet
+     * \param[in] character The character to process. Must model seqan3::semi_alphabet_concept.
+     *
+     * \returns size_t.
+     * \sa seqan3::to_rank.
+     */
+    size_t operator()(alphabet_t const & character) const noexcept
+    {
+        using seqan3::to_rank;
+        return to_rank(character);
+    }
+};
+
+/*!\brief Struct for hashing a range of characters.
+ * \ingroup alphabet
+ * \tparam urng_t The type of the range; Must model std::ranges::InputRange and the value type of the range of the range
+                  must model seqan3::semi_alphabet_concept.
+ */
+template <ranges::InputRange urng_t>
+    //!\cond
+    requires seqan3::semi_alphabet_concept<seqan3::value_type_t<urng_t>>
+    //!\endcond
+struct hash<urng_t>
+{
+    /*!\brief Compute the hash for a range of characters.
+     * \ingroup alphabet
+     * \param[in] range The input range to process. Must model std::ranges::InputRange and the value type of the range
+                        of the range must model seqan3::semi_alphabet_concept.
+     * \returns size_t.
+     */
+    size_t operator()(urng_t const & range) const noexcept
+    {
+        using alphabet_t = seqan3::value_type_t<urng_t>;
+        size_t result{0};
+        hash<alphabet_t> h{};
+        return ::ranges::v3::accumulate(range, result, [&h](size_t const result, alphabet_t const c) {
+            return result * alphabet_t::value_size + h(c);
+        });
```

Comment:
* See previous commen regarding `ranges::`
* `{` to the beginning of the next line, space between `[&h] (size_t)`
* don't use`alphabet_t::value_size`, instead use `seqan3::alphabet_size_v<alphabet_t>`